utf8_to_uv_msgs: Do some code cleanup

khwilliamson opened the pull request:
This should not change behavior.
It moves declarations and initializations closer to first use.  The biggest change is that several cases in a switch have virtually identical logic, which this extracts out into common code

* This set of changes does not require a perldelta entry.

khwilliamson committed:
utf8_to_uv_msgs(): Move dTHX to only block used in

This only is needed if a problem was found.

khwilliamson committed:
utf8_to_uv_msgs: Move declaration to first need

There are several paths through the code that don't need this copy to be
made.  Move to just before it is really needed.

khwilliamson committed:
utf8_to_uv_msgs: Combine duplicate statements

By deferring to the end of the function the setting of a variable
returned to the caller, we can eliminate one copy of that setting.  The
early returns from this function that necessitated the early setting
have been removed in previous commits.

khwilliamson committed:
utf8_to_uv_msgs: Rationalize use of two variables

This makes these two variables always contain the value their names
indicate.

khwilliamson committed:
utf8_to_uv_msgs: Move an assert

This is a small detail, but this moves this assert to after a
conditional that would exclude it.  That is, if the conditional is true,
the assert is pointless.  So move the assert to where we know the
conditional is false.

khwilliamson committed:
utf8_to_uv_msgs: Remove duplicate code

This was introduced in a rebasing error

tonycoz commented on utf8.c:
```diff
@@ -2069,7 +2069,8 @@ Perl_utf8_to_uv_msgs_helper_(const U8 * const s0,
               default:
                 Perl_croak(aTHX_ "panic: Unexpected case value in "
-                                 " utf8n_to_uvchr_msgs() %d", this_problem);
+                                 " utf8n_to_uvchr_msgs() %ld",
+                                 (long) this_problem);
```
It would be more correct to use `U32uf` here

khwilliamson replied:
Done by #22899

khwilliamson committed:
utf8_to_uv_msgs: Move decls and inits closer to first use

C99 allows us to declare anywhere; so move these to where its more
logical.  It also makes sure some variables are initialized before the
goto that jumps to the end of the program, and which currently doesn't
rely on these values, but could be changed to do so someday without the
coder realizing it.  This prevents a problem in case that happens.

khwilliamson commented:
I should have fixed this up long before now.  Some of the commits in the original PR still aren't ready yet, but others were and fixed compiler warnings that others ended up spending duplicate effort in fixing
